contracts/apps/finance/FinanceApp.sol

```diff
@@ -0,0 +1,428 @@
+        uint256 income;
+    }
+
+    struct Period {
+        uint64 startTime;
+        uint64 endTime;
+        uint256 firstTransactionId;
+        uint256 lastTransactionId;
+
+        mapping (address => TokenStatement) tokenStatement;
+    }
+
+    struct Settings {
+        uint64 periodDuration;
+        mapping (address => uint256) budgets;
+    }
+
+    Vault public vault;
+
+    Payment[] payments; // first index is 1
+    Transaction[] transactions; // first index is 1
+    Period[] periods; // first index is 0
+    Settings settings;
+
+    event NewPeriod(uint256 indexed periodId, uint64 periodStarts, uint64 periodEnds);
+    event SetBudget(address indexed token, uint256 amount);
+    event NewPayment(uint256 indexed paymentId, address indexed recipient, uint64 maxRepeats);
+    event NewTransaction(uint256 indexed transactionId, bool incoming, address indexed entity);
+    event ChangePaymentState(uint256 indexed paymentId, bool disabled);
+    event ChangePeriodDuration(uint64 newDuration);
+    event PaymentFailure(uint256 paymentId);
+
+    // Modifier used by all methods that impact accounting to make sure accounting period
+    // is changed before the operation if needed
+    modifier transitionsPeriod {
+        bool completeTransition = tryTransitionAccountingPeriod(MAX_PERIOD_TRANSITIONS_PER_TX);
+        require(completeTransition);
+        _;
+    }
+
+    /**
+    * @notice Initialize Finance app for `_vault` with duration `_periodDuration`
+    * @param _vault Address of the vault Finance will rely on (non changeable)
+    * @param _periodDuration Duration in seconds of each period
+    */
+    function initialize(Vault _vault, uint64 _periodDuration) onlyInit {
+        initialized();
+
+        require(_periodDuration > 1);
+
+        vault = _vault;
+
+        payments.length += 1;
+        payments[0].disabled = true;
+
+        transactions.length += 1;
+
+        settings.periodDuration = _periodDuration;
+        _newPeriod(uint64(getTimestamp()));
+    }
+
+    /**
+    * @dev Deposit for ERC20 approved tokens
+    * @notice Send `_amount` `_token`
+    * @param _token Address of deposited token
+    * @param _amount Amount of tokens sent
+    */
+    function deposit(ERC20 _token, uint256 _amount) transitionsPeriod {
+        _recordTransaction(
+            true, // incoming transaction
+            _token,
+            msg.sender,
+            _amount,
+            0 // unrelated to any existing payment
+        );
+
+        require(_token.transferFrom(msg.sender, address(vault), _amount));
+    }
+
+    function newPayment(
+        ERC20 _token,
+        address _receiver,
+        uint256 _amount,
+        uint64 _initialPaymentTime,
+        uint64 _interval,
+        uint64 _maxRepeats,
+        string _reference
+    ) auth(PAYMENT_CREATOR_ROLE) transitionsPeriod external returns (uint256 paymentId)
+    {
+
+        require(settings.budgets[_token] > 0); // Token must have been added to budget
```

Comment:
What if you have no budgets?

Reply:
You should have a budget in a token before creating a payment that withdraws from that.

There could still be the case in which a token budget that a recurring payment depends on is set to 0, which will make payment executions fail.

Reply:
The question is whether or not we want to enforce users to create budgets. Is it a technical limitation, or is it a UX one? If it's a UX one, then I don't know if it's a good idea.

I mean, budgetting is good, but not everyone does it, as it introduces some overhead.

Reply:
It is a purely UX decision.

I like the idea of allowing not to have budgets and just allow unlimited spending in a given token.  The question is what should be the default behavior. I think it would be a good idea requiring users to explicitly set an 'unlimited' budget in a token for allowing that.

@luisivan thoughts?

Reply:
Talked to luis about this and decided that unless a budget is explicitly set, unlimited spending can happen on that token.

We need to add a new function to free a token from having a budget, as it is a different behavior than setting it to 0.

Reply:
Cool!